akka-stream-tests/src/test/scala/akka/stream/impl/StreamLayoutSpec.scala

```diff
@@ -0,0 +1,240 @@
+  }
+
+  def testStage(): Module = testAtomic(1, 1)
+  def testSource(): Module = testAtomic(0, 1)
+  def testSink(): Module = testAtomic(1, 0)
+
+  val ignore: (Any, Any) ⇒ Any = (x, y) ⇒ ()
+
+  "StreamLayout" must {
+
+    "be able to model simple linear stages" in {
+      val stage1 = testStage()
+
+      stage1.inPorts.size should be(1)
+      stage1.outPorts.size should be(1)
+      stage1.isRunnable should be(false)
+      stage1.isFlow should be(true)
+      stage1.isSink should be(false)
+      stage1.isSource should be(false)
+
+      val stage2 = testStage()
+      val flow12 = stage1.grow(stage2, ignore).connect(stage1.outPorts.head, stage2.inPorts.head)
+
+      flow12.inPorts should be(stage1.inPorts)
+      flow12.outPorts should be(stage2.outPorts)
+      flow12.isRunnable should be(false)
+      flow12.isFlow should be(true)
+      flow12.isSink should be(false)
+      flow12.isSource should be(false)
+
+      val source0 = testSource()
+      source0.inPorts.size should be(0)
+      source0.outPorts.size should be(1)
+      source0.isRunnable should be(false)
+      source0.isFlow should be(false)
+      source0.isSink should be(false)
+      source0.isSource should be(true)
+
+      val sink3 = testSink()
+      sink3.inPorts.size should be(1)
+      sink3.outPorts.size should be(0)
+      sink3.isRunnable should be(false)
+      sink3.isFlow should be(false)
+      sink3.isSink should be(true)
+      sink3.isSource should be(false)
+
+      val source012 = source0.grow(flow12, ignore).connect(source0.outPorts.head, flow12.inPorts.head)
+      source012.inPorts.size should be(0)
+      source012.outPorts should be(flow12.outPorts)
+      source012.isRunnable should be(false)
+      source012.isFlow should be(false)
+      source012.isSink should be(false)
+      source012.isSource should be(true)
+
+      val sink123 = flow12.grow(sink3, ignore).connect(flow12.outPorts.head, sink3.inPorts.head)
+      sink123.inPorts should be(flow12.inPorts)
+      sink123.outPorts.size should be(0)
+      sink123.isRunnable should be(false)
+      sink123.isFlow should be(false)
+      sink123.isSink should be(true)
+      sink123.isSource should be(false)
+
+      val runnable0123a = source0.grow(sink123, ignore).connect(source0.outPorts.head, sink123.inPorts.head)
+      val runnable0123b = source012.grow(sink3, ignore).connect(source012.outPorts.head, sink3.inPorts.head)
+
+      val runnable0123c =
+        source0
+          .grow(flow12, ignore).connect(source0.outPorts.head, flow12.inPorts.head)
+          .grow(sink3, ignore).connect(flow12.outPorts.head, sink3.inPorts.head)
+
+      runnable0123a.inPorts.size should be(0)
+      runnable0123a.outPorts.size should be(0)
+      runnable0123a.isRunnable should be(true)
+      runnable0123a.isFlow should be(false)
+      runnable0123a.isSink should be(false)
+      runnable0123a.isSource should be(false)
+    }
+
+    "be able to model hierarchic linear modules" in {
+      pending
```

Comment:
I won't fix these now, but we need more tests. It is a huge benefit of the new materialization system that the logic can be tested in isolation. We should use this power. (also, this is one lession from remoting)